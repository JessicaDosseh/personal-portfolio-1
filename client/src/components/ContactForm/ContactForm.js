import React, { useState, useEffect } from "react";
import PropTypes from "prop-types";
import styled, { css, keyframes } from "styled-components";
import axios from "axios";
import validator from "email-validator";
import { useStateValue } from "react-conflux";
import { formContext } from "../../store/contexts";
import { HANDLE_FORM_CHANGE } from "../../store/reducers/formReducer";

const serverURL =
  process.env.SERVER_URL || "https://nathan-portfolio-backend.herokuapp.com/";

const buttonHover = keyframes`
  0% {
    opacity: 1;
  }
  100% {
    opacity: 0.8;
  }
`;

const buttonHoverDark = keyframes`
  0% {
    opacity: 1;
  }
  100% {
    opacity: 0.6;
  }
`;

const ContactFormContainer = styled.div`
  margin: 50px auto 0;
  width: 100%;
  max-width: 600px;

  ${props =>
    props.darkmode &&
    css`
      color: white;
    `}

  h1 {
    width: 100%;
    text-align: center;
    margin-bottom: 35px;
    font-size: 3rem;
  }

  form {
    display: flex;
    flex-direction: column;

    label {
      margin-bottom: 10px;
    }

    span {
      margin-bottom: 10px;
    }

    input {
      margin-bottom: 30px;
      height: 45px;
      border: 1px solid #2f3234;
      border-radius: 5px;
      -moz-border-radius: 5px;
      -webkit-border-radius: 5px;
      padding: 5px;
      font-size: 1.6rem;
      font-family: "Work Sans", sans-serif;
      background: #f7f7f7;

      ${props =>
        props.darkmode &&
        css`
          background: white;
        `}

      &:last-of-type {
        display: none !important;
      }
    }

    textarea {
      background: #f7f7f7;
      border: 1px solid #2f3234;
      width: 100%;
      padding: 5px;
      font-size: 1.6rem;
      border-radius: 5px;
      -moz-border-radius: 5px;
      -webkit-border-radius: 5px;

      ${props =>
        props.darkmode &&
        css`
          background: white;
        `}
    }

    div {
      display: flex;
      flex-direction: column;
      padding-bottom: 100px;

      @media (min-width: 500px) {
        flex-direction: row;
        justify-content: space-between;
      }

      button {
        width: 100%;
        height: 45px;
        margin: 30px 0 0;
        border: 0;
        border-radius: 5px;
        -moz-border-radius: 5px;
        -webkit-border-radius: 5px;
        background: #ad91ed;
        /* #b6465f */
        color: white;
        font-size: 1.6rem;
        text-transform: capitalize;
        cursor: pointer;

        @media (min-width: 500px) {
          width: 47%;
        }

        ${props =>
          props.darkmode &&
          css`
            background: black;
          `}

        &:hover {
          animation: ${buttonHover} 0.2s forwards;

          ${props =>
            props.darkmode &&
            css`
              animation: ${buttonHoverDark} 0.2s forwards;
            `}
        }
      }
    }
  }
`;

const EmailValidation = styled.p`
  display: inline-block;
  font-size: 1.6rem;
  font-weight: 200;
  font-family: "Work Sans", sans-serif;
  color: #f24b6c;
  margin-left: 2%;
  line-height: 1;
`;

const ContactForm = props => {
  const [state, dispatch] = useStateValue(formContext);
  const {
    contactName,
    contactEmail,
    contactSubject,
    contactMessage,
    contact_me_by_fax_only
  } = state;
  const [isEmail, setIsEmail] = useState(null);
  useEffect(() => {
    window.scrollTo(0, 0);
  }, []);
  const validateEmail = emailValue => {
    // Validates email or sets value of hook isEmail to null if empty string
    if (!emailValue) {
      setIsEmail(null);
    } else {
      let validEmail = null;
      validEmail = validator.validate(emailValue);
      setIsEmail(validEmail);
    }
  };
  const handleChange = e => {
    e.preventDefault();
    if (e.target.name === "contactEmail") validateEmail(e.target.value);
    dispatch({
      type: HANDLE_FORM_CHANGE,
      payload: { name: e.target.name, value: e.target.value }
    });
  };
  const cancelForm = e => {
    props.history.push("/");
  };
  const sendEmail = e => {
    e.preventDefault();
    if (!isEmail) {
      return props.setInvalidEmailModal(true);
    } else {
      // Send email to server
      props.setEmailValues({
        ...props.emailValues,
        sending: true
      });
      const email = {
        name: contactName,
        email: contactEmail,
        subject: contactSubject,
        message: contactMessage,
        honeyField: contact_me_by_fax_only
      };
      axios
        .post(serverURL, email)
        .then(res => {
          window.scrollTo(0, 0);
          if (res.statusText === "OK") {
            props.setEmailValues({
              sending: false,
              success: true,
              displayModal: true,
              modalText: "Message sent successfully."
            });
          } else {
            props.setEmailValues({
              sending: false,
              success: false,
              displayModal: false,
              modalText: "Message did not go through."
            });
          }
        })
        .catch(err => {
          window.scrollTo(0, 0);
          props.setEmailValues({
            sending: false,
            success: false,
            displayModal: false,
            modalText: "Message did not go through."
          });
        });
    }
  };
  return (
    <ContactFormContainer darkmode={props.darkmode}>
      <h1>Get in Contact</h1>
      <form onSubmit={sendEmail}>
        <label htmlFor="name">Name</label>
        <input
          data-testid="nameInput"
          id="name"
          required
          autoComplete="off"
          type="text"
          name="contactName"
          value={contactName}
          onChange={handleChange}
        />
        <span>
          <label htmlFor="email">Email</label>
          {!isEmail && isEmail !== null && (
            <EmailValidation>(Invalid email)</EmailValidation>
          )}
        </span>
        <input
          data-testid="emailInput"
          id="email"
          required
          autoComplete="off"
          type="text"
          name="contactEmail"
          value={contactEmail}
          onChange={handleChange}
        />
        <label htmlFor="subject">Subject</label>
        <input
          data-testid="subjectInput"
          id="subject"
          required
          autoComplete="off"
          type="text"
          name="contactSubject"
          value={contactSubject}
          onChange={handleChange}
        />
        <label htmlFor="message">Message</label>
        <textarea
          data-testid="contactInput"
          id="message"
          required
          autoComplete="off"
          rows="12"
          name="contactMessage"
          value={contactMessage}
          onChange={handleChange}
        />
        <input
          id="fax"
          type="text"
          name="contact_me_by_fax_only"
          value={contact_me_by_fax_only}
          onChange={handleChange}
          tabIndex="-1"
          autoComplete="off"
        />
        <div>
          <button type="submit">Submit</button>
          <button type="button" onClick={cancelForm}>
            Cancel
          </button>
        </div>
      </form>
    </ContactFormContainer>
  );
};

ContactForm.propTypes = {
  darkmode: PropTypes.bool.isRequired,
  value: PropTypes.arrayOf(
    PropTypes.shape({
      contactName: PropTypes.string,
      contactEmail: PropTypes.string,
      contactSubject: PropTypes.string,
      contactMessage: PropTypes.string
    })
  ),
  setInvalidEmailModal: PropTypes.func
};

export default ContactForm;
