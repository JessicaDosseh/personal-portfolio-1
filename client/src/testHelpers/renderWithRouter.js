import React from "react";
import { Router } from "react-router-dom";
import { render } from "@testing-library/react";
import { createMemoryHistory } from "history";
import { StateProvider } from "react-conflux";
import { globalReducer, globalContext } from "../store/reducers/globalReducer";

const renderWithRouter = (
  ui,
  {
    route = "/",
    history = createMemoryHistory({ initialEntries: [route] })
  } = {}
) => {
  return {
    ...render(
      <StateProvider reducer={globalReducer} stateContext={globalContext}>
        <Router history={history}>{ui}</Router>
      </StateProvider>
    ),
    history
  };
};

export default renderWithRouter;
