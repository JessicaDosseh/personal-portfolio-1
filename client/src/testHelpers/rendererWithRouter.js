import React from "react";
import { BrowserRouter as Router } from "react-router-dom";
import renderer from "react-test-renderer";
import { createMemoryHistory } from "history";
import { StateProvider } from "react-conflux";
import { globalReducer, globalContext } from "../store/reducers/globalReducer";

const rendererWithRouter = (
  ui,
  {
    route = "/",
    history = createMemoryHistory({ initialEntries: [route] })
  } = {}
) => {
  return {
    ...renderer.create(
      <StateProvider reducer={globalReducer} stateContext={globalContext}>
        <Router history={history}>{ui}</Router>
      </StateProvider>
    ),
    history
  };
};

export default rendererWithRouter;
