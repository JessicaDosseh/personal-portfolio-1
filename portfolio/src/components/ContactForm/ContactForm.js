import React from "react";
import PropTypes from "prop-types";
import styled, { css } from "styled-components";

const ContactFormContainer = styled.div`
  margin: 50px 5%;
  width: 90%;

  ${props =>
    props.darkmode &&
    css`
      color: white;
    `}

  form {
    display: flex;
    flex-direction: column;

    p {
      margin-bottom: 15px;
    }

    input {
      margin-bottom: 15px;
      height: 40px;
      border: 1px solid #2f3234;
      border-radius: 3px;
      padding: 5px;
      font-size: 1.4rem;
      font-family: "Work Sans", sans-serif;
    }
  }
`;

const ContactForm = props => {
  return (
    <ContactFormContainer darkmode={props.darkmode}>
      <form>
        <p>Your Name</p>
        <input required type="text" name="contactName" placeholder="Name" />
        <p>Your Email</p>
        <input required type="text" name="contactEmail" placeholder="Email" />
        <p>Subject</p>
        <input
          required
          type="text"
          name="contactSubject"
          placeholder="Subject"
        />
        <p>Your Message</p>
        <textarea required rows="10" name="contactSubject" />
        <button type="submit">SUBMIT</button>
        <button type="button">CLEAR</button>
      </form>
    </ContactFormContainer>
  );
};

ContactForm.propTypes = {};

export default ContactForm;
