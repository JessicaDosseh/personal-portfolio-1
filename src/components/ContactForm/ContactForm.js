import React, { useState } from "react";
import PropTypes from "prop-types";
import styled, { css, keyframes } from "styled-components";
import axios from "axios";

const buttonHover = keyframes`
  0% {
    opacity: 1;
  }
  100% {
    opacity: 0.8;
  }
`;

const buttonHoverDark = keyframes`
  0% {
    background: black;
  }
  100% {
    background: #030303;
  }
`;

const ContactFormContainer = styled.div`
  margin: 50px auto 0;
  width: 100%;
  max-width: 1200px;

  ${props =>
    props.darkmode &&
    css`
      color: white;
    `}

  form {
    display: flex;
    flex-direction: column;

    h4 {
      margin-bottom: 10px;
    }

    input {
      margin-bottom: 30px;
      height: 45px;
      border: 1px solid #2f3234;
      padding: 5px;
      font-size: 1.6rem;
      font-family: "Work Sans", sans-serif;
      background: #f7f7f7;

      ${props =>
        props.darkmode &&
        css`
          background: white;
        `}
    }

    textarea {
      background: #f7f7f7;
      border: 1px solid #2f3234;
      width: 100%;
      padding: 5px;
      font-size: 1.6rem;

      ${props =>
        props.darkmode &&
        css`
          background: white;
        `}
    }

    div {
      display: flex;
      flex-direction: column;
      margin-bottom: 120px;

      @media (min-width: 500px) {
        flex-direction: row;
        justify-content: space-between;
      }

      button {
        width: 100%;
        height: 45px;
        margin: 30px 0 0;
        border: 0;
        border-radius: 8px;
        -moz-border-radius: 8px;
        -webkit-border-radius: 8px;
        background: #ad91ed;
        /* #b6465f */
        color: white;
        font-size: 1.6rem;
        text-transform: capitalize;
        cursor: pointer;

        @media (min-width: 500px) {
          width: 47%;
        }

        ${props =>
          props.darkmode &&
          css`
            background: black;
          `}

        &:hover {
          animation: ${buttonHover} 0.2s forwards;

          ${props =>
            props.darkmode &&
            css`
              animation: ${buttonHoverDark} 0.2s forwards;
            `}
        }
      }
    }
  }
`;

const ContactForm = props => {
  (function onLoadSetTop() {
    window.scrollTo(0, 0);
  })();
  const [value, setValue] = useState({
    contactName: "",
    contactEmail: "",
    contactSubject: "",
    contactMessage: ""
  });
  const handleChange = e => {
    e.preventDefault();
    setValue({
      ...value,
      [e.target.name]: e.target.value
    });
  };
  const clearForm = e => {
    e.preventDefault();
    setValue({
      contactName: "",
      contactEmail: "",
      contactSubject: "",
      contactMessage: ""
    });
  };
  const sendEmail = e => {
    e.preventDefault();
    const email = {
      name: value.contactName,
      email: value.contactEmail,
      subject: value.contactSubject,
      message: value.contactMessage
    };
    axios
      .post("http://localhost:5000/", email)
      .then(res => console.log(res))
      .catch(err => console.log(err));
  };
  return (
    <ContactFormContainer darkmode={props.darkmode}>
      <form onSubmit={sendEmail}>
        <h4>Name</h4>
        <input
          required
          autoComplete="off"
          type="text"
          name="contactName"
          placeholder="Name"
          value={value.contactName}
          onChange={handleChange}
        />
        <h4>Email</h4>
        <input
          required
          autoComplete="off"
          type="text"
          name="contactEmail"
          placeholder="Email"
          value={value.contactEmail}
          onChange={handleChange}
        />
        <h4>Subject</h4>
        <input
          required
          autoComplete="off"
          type="text"
          name="contactSubject"
          placeholder="Subject"
          value={value.contactSubject}
          onChange={handleChange}
        />
        <h4>Message</h4>
        <textarea
          required
          autoComplete="off"
          rows="12"
          name="contactMessage"
          value={value.contactMessage}
          onChange={handleChange}
        />
        <div>
          <button type="submit">Submit</button>
          <button type="button" onClick={clearForm}>
            Clear
          </button>
        </div>
      </form>
    </ContactFormContainer>
  );
};

// ContactForm.propTypes = {
//   darkmode: PropTypes.bool.isRequired,
//   value: PropTypes.arrayOf({
//     contactName: PropTypes.string,
//     contactEmail: PropTypes.string,
//     contactSubject: PropTypes.string,
//     contactMessage: PropTypes.string
//   })
// };

ContactForm.defaultProps = {
  value: PropTypes.shape(
    PropTypes.shape({
      contactName: "",
      contactEmail: "",
      contactSubject: "",
      contactMessage: ""
    })
  )
};

export default ContactForm;
