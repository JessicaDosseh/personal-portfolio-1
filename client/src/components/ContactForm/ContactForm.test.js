import React from "react";
import "react-testing-library/cleanup-after-each";
import "jest-dom/extend-expect";
import { renderWithRouter, rendererWithRouter } from "../../testHelpers";

import ContactForm from "./ContactForm";

describe("<ContactForm />", () => {
  it("should render the ContactForm component without crashing", () => {
    const push = jest.fn();
    const scrollToSpy = jest.fn();
    window.scrollTo = scrollToSpy;
    const helpers = renderWithRouter(
      <ContactForm darkmode={false} history={push} />
    );
  });

  it("should match the snapshot of ContactForm", () => {
    const tree = rendererWithRouter(<ContactForm darkmode={false} />);
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it("should render the ContactForm text", () => {
    const { getByText } = renderWithRouter(<ContactForm darkmode={false} />);
    expect(getByText(/Get in Contact/i)).toBeDefined();
  });

  it("should render the inputs for the contact form", () => {
    const { getByTestId } = renderWithRouter(<ContactForm darkmode={false} />);
    expect(getByTestId("nameInput")).toBeDefined();
    expect(getByTestId("emailInput")).toBeDefined();
    expect(getByTestId("subjectInput")).toBeDefined();
    expect(getByTestId("contactInput")).toBeDefined();
  });
});
