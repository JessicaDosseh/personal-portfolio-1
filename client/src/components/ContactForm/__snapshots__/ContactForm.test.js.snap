// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ContactForm /> should match the snapshot of ContactForm 1`] = `
<div
  className="sc-AykKC clyZZx"
>
  <h1>
    Get in Contact
  </h1>
  <form
    onSubmit={[Function]}
  >
    <label
      htmlFor="name"
    >
      Name
    </label>
    <input
      autoComplete="off"
      data-testid="nameInput"
      id="name"
      name="contactName"
      onChange={[Function]}
      required={true}
      type="text"
      value=""
    />
    <span>
      <label
        htmlFor="email"
      >
        Email
      </label>
    </span>
    <input
      autoComplete="off"
      data-testid="emailInput"
      id="email"
      name="contactEmail"
      onChange={[Function]}
      required={true}
      type="text"
      value=""
    />
    <label
      htmlFor="subject"
    >
      Subject
    </label>
    <input
      autoComplete="off"
      data-testid="subjectInput"
      id="subject"
      name="contactSubject"
      onChange={[Function]}
      required={true}
      type="text"
      value=""
    />
    <label
      htmlFor="message"
    >
      Message
    </label>
    <textarea
      autoComplete="off"
      data-testid="contactInput"
      id="message"
      name="contactMessage"
      onChange={[Function]}
      required={true}
      rows="12"
      value=""
    />
    <input
      autoComplete="off"
      id="fax"
      name="contact_me_by_fax_only"
      onChange={[Function]}
      tabIndex="-1"
      type="text"
      value=""
    />
    <div>
      <button
        type="submit"
      >
        Submit
      </button>
      <button
        onClick={[Function]}
        type="button"
      >
        Cancel
      </button>
    </div>
  </form>
</div>
`;
